File: frontend/src/api/apertura.ts
import { apiFetch } from './client'

/**
 * Apertura del sistema: los saldos con los que el negocio arrancó a usarlo.
 *
 * Cuando el sistema se puso en marcha el negocio ya venía funcionando: había
 * efectivo en el cajón y cheques en cartera comprados tiempo atrás. Los dos son
 * saldos de apertura, no operaciones del día.
 */
export interface ConfiguracionApertura {
  /** Hasta este día inclusive, los cheques cargados son cartera preexistente. */
  fecha_corte_carga_inicial: string | null
  saldo_inicial_ars: string | null
  saldo_inicial_usd: string | null
  /** Día al que corresponde el efectivo, no el día en que se cargó. */
  fecha_saldo_inicial: string | null
  definido_por: string | null
  definido_at: string | null
  saldo_definido: boolean
}

export interface FechaCorteResponse {
  fecha_corte: string
  cheques_marcados: number
  lineas_revertidas: number
}

export const getApertura = (): Promise<ConfiguracionApertura> =>
  apiFetch<ConfiguracionApertura>('/apertura')

export const definirFechaCorte = (payload: {
  fecha_corte: string
  operador_id: string
}): Promise<FechaCorteResponse> =>
  apiFetch<FechaCorteResponse>('/apertura/fecha-corte', {
    method: 'POST',
    body: JSON.stringify(payload),
  })

export const definirSaldoInicial = (payload: {
  saldo_ars: number
  saldo_usd: number
  fecha: string
  operador_id: string
  forzar?: boolean
}): Promise<ConfiguracionApertura> =>
  apiFetch<ConfiguracionApertura>('/apertura/saldo-inicial', {
    method: 'POST',
    body: JSON.stringify(payload),
  })

File: backend/app/api/routes/apertura.py
"""Rutas de apertura del sistema — los saldos con los que arrancó el negocio.

Dos operaciones de puesta en marcha: fijar hasta cuándo los cheques que se cargan
son cartera preexistente, y cargar el efectivo que había en el cajón.
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.schemas.apertura import (
    ConfiguracionAperturaRead,
    FechaCorteRequest,
    FechaCorteResponse,
    SaldoInicialRequest,
)
from app.services import apertura as service

router = APIRouter(prefix="/apertura", tags=["apertura"])

DbSession = Annotated[Session, Depends(get_db)]


@router.get("", response_model=ConfiguracionAperturaRead)
def get_apertura(db: DbSession) -> ConfiguracionAperturaRead:
    """Estado actual: qué falta definir y qué ya quedó fijado."""
    return service.get_configuracion(db)


@router.post("/fecha-corte", response_model=FechaCorteResponse)
def definir_fecha_corte(payload: FechaCorteRequest, db: DbSession) -> FechaCorteResponse:
    """Fija hasta qué día los cheques cargados son cartera preexistente.

    Corrige además hacia atrás: a los cheques ya cargados dentro del período les
    quita el egreso de compra que no correspondía."""
    resultado = service.definir_fecha_corte(
        db, payload.fecha_corte, operador_id=payload.operador_id
    )
    return FechaCorteResponse(fecha_corte=payload.fecha_corte, **resultado)


@router.post("/saldo-inicial", response_model=ConfiguracionAperturaRead)
def definir_saldo_inicial(
    payload: SaldoInicialRequest, db: DbSession
) -> ConfiguracionAperturaRead:
    """Carga el efectivo de arranque, por moneda. Por única vez."""
    return service.definir_saldo_inicial(
        db,
        saldo_ars=payload.saldo_ars,
        saldo_usd=payload.saldo_usd,
        cotizacion_usd=payload.cotizacion_usd,
        fecha=payload.fecha,
        operador_id=payload.operador_id,
        forzar=payload.forzar,
    )
